### PR TITLE
Refactor data.io

### PR DESCRIPTION
**Pull request by astaric**
Current singleton implementation of file readers/writers is hard to extend when additional features are needed (such as selecting sheet for excel files and overwriting header in #1012).

This PR replaces
    FileFormat.read(filename)
with
    FileFormat.get_reader(filename).read()

The new way enables us to set additional properties on reader before the read method is called.

### CODE REVIEW

File: Orange/data/io.py [hunk 1]
```diff
@@ -626,43 +651,46 @@ class BasketFormat(FileFormat):
...
 
-    @classmethod
-    def read_file(cls, filename, storage_class=None):
-        from Orange.data import _io, Table, Domain
-        import sys
-        if storage_class is None:
-            storage_class = Table
+    def read(self):
+        if self.wrapper is None or self.wrapper is _IDENTITY:
+            self.wrapper = Table
```
Comment (kernc) [inline Orange/data/io.py]: I think we should consider throwing these wrappers out. Is `wrapper` ever not `Table`? If it isn't then the caller is obviously aware of it and can invoke `wrapper(reader.read())` on their end.

File: Orange/data/table.py [hunk 2]
```diff
@@ -517,7 +517,10 @@ def from_file(cls, filename, wrapper=None):
 
         if not os.path.exists(absolute_filename):
             raise IOError('File "{}" was not found.'.format(filename))
-        data = FileFormat.read(absolute_filename, wrapper or cls)
+        reader = FileFormat.get_reader(absolute_filename)
+        if wrapper or cls != Table:
+            reader.set_wrapper(wrapper or cls)
+        data = reader.read()
```
Comment (kernc) [inline Orange/data/table.py]: Aware ... like ... here.

File: Orange/data/table.py [hunk 3]
```diff
@@ -500,24 +500,12 @@ def from_file(cls, filename, wrapper=None):
...
-            absolute_filename = ext = ""
 
-        if not os.path.exists(absolute_filename):
-            raise IOError('File "{}" was not found.'.format(filename))
-        data = FileFormat.read(absolute_filename, wrapper or cls)
+        absolute_filename = FileFormat.locate(filename, dataset_dirs)
+        reader = FileFormat.get_reader(absolute_filename)
+        if wrapper or cls != Table:
+            reader.set_wrapper(wrapper or cls)
+        data = reader.read()
```
Comment (kernc) [inline Orange/data/table.py]: Aware ... like ... here.

File: Orange/data/io.py (hunk 1 above)
Comment (astaric) [inline Orange/data/io.py]: I agree. See ce70666

**Pull request merged**